### PR TITLE
[Feature] Adds admin GUI

### PR DESCRIPTION
**Pull request by Sqyphen**
## Description

Adds an admin GUI with a form for creating new debates
There is no persistent data storage yet so the form submission is a dummy api call (which has also been created)

## Implementation

+ Adds admin gui front end
+ adds s3o 
+ restructures routing
+ adds skeleton api route

## Screenshot

Alt text

## .env changes

+ Copy the _.env-example_ file to create a new .env (`cp .env-example .env`)
+ Update the .env file with the PORT number you'd like to use

## Additional requirements

Please run `npm install` again on switching to this branch, multiple packages have been updated

Alt text

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2,14 +2,32 @@ const dotenv = require("dotenv").config({
   silent: process.env.NODE_ENVIRONMENT === "production"
 });
 
+global.__basedir = __dirname;
```
Comment (Lily2point0) [inline index.js]: Use `path` instead.

File: index.js [hunk 2]
```diff
@@ -20,27 +38,36 @@ app.engine(
...
+let requestLogger = function(req, res, next) {
+  debug("RECEIVED REQUEST:", req.method, req.url);
+  next();
+};
 
-const PORT = process.env.PORT;
-if (!PORT) {
-  throw new Error("ERROR: PORT not specified in env");
-}
+hbs.registerHelper("greaterThan", function(value1, value2) {
```
Comment (Lily2point0) [inline index.js]: Is this required? Could hbs helpers be kept in a self-contained file?

Comment (Sqyphen) [reply index.js]: Ah, I've removed this - it wasn't working anyway

File: index.js [hunk 3]
```diff
@@ -2,14 +2,32 @@ const dotenv = require("dotenv").config({
...
+const admin_routes = require("./routes/admin");
+const debate_routes = require("./routes/debate");
+const api_routes = require("./routes/api");
+
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(bodyParser.json());
 
-// view engine setup
-var hbs = require("express-hbs");
+if (process.env.NODE_ENV === "production") {
```
Comment (Lily2point0) [inline index.js]: You'll want to enforce ssl before enabling `bodyParser`

File: index.js [hunk 4]
```diff
@@ -2,14 +2,32 @@ const dotenv = require("dotenv").config({
   silent: process.env.NODE_ENVIRONMENT === "production"
```
Comment (Lily2point0) [inline index.js]: The correct var for Heroku is `NODE_ENV`

File: index.js (hunk 1 above)
Comment (Sqyphen) [reply index.js]: Done and removed

File: index.js (hunk 4 above)
Comment (Sqyphen) [reply index.js]: Updated, although this did come from the template

File: package.json [hunk 5]
```diff
@@ -4,25 +4,35 @@
...
+    "ftlabs"
+  ],
+  "author": "ftlabs",
+  "license": "MIT",
   "bugs": {
     "url": "https://github.com/ftlabs/order-order/issues"
   },
   "homepage": "https://github.com/ftlabs/order-order#readme",
   "dependencies": {
+    "@financial-times/s3o-middleware": "^2.0.8",
```
Comment (Lily2point0) [inline package.json]: latest is `3.0.1`, fixes s3o PORT issues

File: index.js (hunk 3 above)
Comment (Sqyphen) [reply index.js]: Updated

File: package.json (hunk 5 above)
Comment (Sqyphen) [reply package.json]: Nice, thank you

File: static/js/hats/main.js [hunk 6]
```diff
@@ -1,30 +1,45 @@
+const originCommentUsername = document.getElementById("comment-new-username");
```
Comment (Lily2point0) [inline static/js/hats/main.js]: avoid const/let on client-side js for cross-browser compatibility. We don't have a build process to take care of this.

File: static/js/hats/main.js [hunk 7]
```diff
@@ -1,30 +1,45 @@
...
 const originCommentText = document.getElementById("comment-new-text");
 const originCommentOpt = document.getElementById("comment-new-hat-opt");
 const originCommentSubmit = document.getElementById("comment-new-submit");
 const commentsList = document.getElementsByClassName("comments_list")[0];
 
+const formRequestAccess = document.getElementById("request_access");
+const btnRequestAccess = document.getElementById("btn_request_access");
+
 function init() {
```
Comment (Lily2point0) [inline static/js/hats/main.js]: Is this function necessary? Can't `addListeners()` be called straight away?

File: static/js/admin/main.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  if (btnSubmit) {
+    addListeners();
+  }
+}
+
+function addListeners() {
+  btnSubmit.addEventListener("click", submitForm);
+}
+
+async function submitForm(e) {
```
Comment (Lily2point0) [inline static/js/admin/main.js]: no `async` support in any version of IE.

File: static/js/hats/main.js (hunk 6 above)
Comment (Sqyphen) [reply static/js/hats/main.js]: Fixed

File: static/js/hats/main.js (hunk 7 above)
Comment (Sqyphen) [reply static/js/hats/main.js]: Sure, to keep the init function structure I pulled the contents of addListeners into the init instead

File: static/js/admin/main.js (hunk 8 above)
Comment (Sqyphen) [reply static/js/admin/main.js]: Fixed, gaveup with XMLHttpRequest and put Polyfill.io in instead
removed the async await and added a promise instead

File: index.js (hunk 4 above)
Comment (Lily2point0) [reply index.js]: Then I shall update the template

File: views/admin/create_new_debate.hbs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+			<label for="debateDescription">Description</label>
+			<input type="text" name="debateDescription" placeholder="Description" required />
+		</fieldset>
+		<label for="series_order">Series Order*</label>
+		<select name="series_order" required>
+			<option selected>1</option>
+			<option>2</option>
+			<option>3</option>
+		</select>
+		<span>*If this debate is part of a series, select the order it should appear in</span>
```
Comment (Lily2point0) [inline views/admin/create_new_debate.hbs]: Would be better for accessibility to have this under the label, and when adding Origami, using `o-forms__additional-info`. Can be left as-is for now.

Comment (Sqyphen) [reply views/admin/create_new_debate.hbs]: Done 😎

**Pull request merged**